pkg/controller/cluster/agent/ports.go

```diff
@@ -0,0 +1,253 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"gopkg.in/yaml.v2"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/registry/core/service/portallocator"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	kubeletPortRangeConfigMapName = "k3k-kubelet-port-range"
+	webhookPortRangeConfigMapName = "k3k-webhook-port-range"
+)
+
+type PortAllocator struct {
+	KubeletCM        *v1.ConfigMap
+	WebhookCM        *v1.ConfigMap
+	KubeletPortRange string
+	WebhookPortRange string
+}
+
+func NewPortAllocator(ctx context.Context, client ctrlruntimeclient.Client, kubeletPortRange, webhookPortRange string) (*PortAllocator, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("starting port allocator")
+
+	var kubeletPortRangeCM, webhookPortRangeCM v1.ConfigMap
+
+	portRangeConfigMapNamespace := os.Getenv("CONTROLLER_NAMESPACE")
+	if portRangeConfigMapNamespace == "" {
+		return nil, fmt.Errorf("failed to find k3k controller namespace")
+	}
+
+	kubeletPortRangeCM.Name = kubeletPortRangeConfigMapName
+	kubeletPortRangeCM.Namespace = portRangeConfigMapNamespace
+
+	webhookPortRangeCM.Name = webhookPortRangeConfigMapName
+	webhookPortRangeCM.Namespace = portRangeConfigMapNamespace
+
+	return &PortAllocator{
+		KubeletCM:        &kubeletPortRangeCM,
+		WebhookCM:        &webhookPortRangeCM,
+		KubeletPortRange: kubeletPortRange,
+		WebhookPortRange: webhookPortRange,
+	}, nil
+}
+
+func (a *PortAllocator) InitPortAllocatorConfig(ctx context.Context, client ctrlruntimeclient.Client) manager.Runnable {
+	return manager.RunnableFunc(func(ctx context.Context) error {
+		if err := a.getOrCreate(ctx, client, a.KubeletCM, a.KubeletPortRange); err != nil {
+			return err
+		}
+
+		if err := a.getOrCreate(ctx, client, a.WebhookCM, a.WebhookPortRange); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
```

Comment:
Which is the reason of having this as a `manager.Runnable`? Can we just initialize it during its creation, in the `NewPortAllocator`?